h2o-automl/src/test/java/ai/h2o/automl/AutoMLTest.java

```diff
@@ -37,26 +41,60 @@
     }
   }
 
-  @Test public void KeepCrossValidationFoldAssignmentTest() {
-    AutoML aml=null;
-    Frame fr=null;
+  @Test public void KeepCrossValidationFoldAssignmentEnabledTest() {
+    AutoML aml = null;
+    Frame fr = null;
+    Model leader = null;
     try {
       AutoMLBuildSpec autoMLBuildSpec = new AutoMLBuildSpec();
       fr = parse_test_file("./smalldata/airlines/allyears2k_headers.zip");
       autoMLBuildSpec.input_spec.training_frame = fr._key;
       autoMLBuildSpec.input_spec.response_column = "IsDepDelayed";
-      autoMLBuildSpec.build_control.stopping_criteria.set_max_runtime_secs(5);
+      autoMLBuildSpec.build_control.stopping_criteria.set_max_models(1);
+      autoMLBuildSpec.build_control.stopping_criteria.set_max_runtime_secs(30);
       autoMLBuildSpec.build_control.keep_cross_validation_fold_assignment = true;
 
       aml = AutoML.makeAutoML(Key.<AutoML>make(), new Date(), autoMLBuildSpec);
       AutoML.startAutoML(aml);
       aml.get();
 
-      assertTrue(aml.leader() !=null && aml.leader()._parms._keep_cross_validation_fold_assignment);
-      assertTrue(aml.leader() !=null && aml.leader()._output._cross_validation_fold_assignment_frame_id != null);
+      leader = aml.leader();
+
+      assertTrue(leader !=null && leader._parms._keep_cross_validation_fold_assignment);
+      assertNotNull(leader._output._cross_validation_fold_assignment_frame_id);
+
+    } finally {
+      // Since user asked to keep cv fold assignments( we set parameter `keep_cross_validation_fold_assignment` to true) we need to remove this key manually
+      Frame cvFoldAssignmentFrame = DKV.getGet(leader._output._cross_validation_fold_assignment_frame_id);
+      cvFoldAssignmentFrame.delete();
```

Comment:
you're right in theory: maybe it would be better to have some kind of `CVSupportModel` abstraction, but currently the cv flags and the cv results are both available in `Model` so I think it's fine to do the cleanup there.
Indeed, as you mentioned, the CV logic is fully handled in `ModelBuilder` which also orchestrates all the models being built, even GridSearch ends up calling ModelBuilder to create new models, but the ModelBuilder doesn't keep any state after this.

Trying to answer your "why wasn't it removed before?" interrogation. I see 2 probable explanations:
- we didn't worry too much about CV data in memory before AutoML because the latter can quickly create a lot of them and waste memory to the point where it prevents AutoML from completing.
- clients that were writing scripts on top of H2O got used to delete those data manually, or.. they didn't delete them at all.

I think that after we call `delete()/remove()` on a model, then every object that is strongly associated with the model, especially everything that was used to build the model and makes sense only in the context of the model, should also be deleted.
So this includes all the temporary frames/vecs, the result metrics and so on.
Somehow, all the objects that are also saved when exporting a model...

Reply:
I was not arguing regarding placing all the cleaning logic into `delete() / remove()`.  I'm just trying to understand the design and assumptions behind it.  Yes, I would expect to have extra abstraction on top of Model because we have some methods that are CV specific in Model.java
```
  public void deleteCrossValidationModels() {
    if (_output._cross_validation_models != null) {
      Log.info("Cleaning up CV Models for " + this._key.toString());
      int count = deleteAll(_output._cross_validation_models);
      Log.info(count+" CV models were removed");
    }
  }

  public void deleteCrossValidationPreds() {
    if (_output._cross_validation_predictions != null) {
      Log.info("Cleaning up CV Predictions for " + this._key.toString());
      int count = deleteAll(_output._cross_validation_predictions);
      Log.info(count+" CV predictions were removed");
    }

    if (_output._cross_validation_holdout_predictions_frame_id != null) {
      _output._cross_validation_holdout_predictions_frame_id.remove();
    }
  }
```
And you see, there are checks inside these methods similar to what you have suggested to put inside Model.remove_impl. So I guess people probably was trying to avoid checks that are unnecessary in all use cases.  My understanding is that we need to fix it....  If all types of models have cross validation support than we need just put checks like -> 
```
if (_output._cross_validation_holdout_predictions_frame_id != null) { 
```
into `Model.remove_impl()` and do cleanup from there. We can remove those two methods/or make them private in Model's API `deleteCrossValidationPreds()` and `deleteCrossValidationModels()`. Also we should remove `cleanUpModelsCVPreds()` method from AutoML.java which is by the way is being called from the end of learn() method and move it into remove pipeline.
 Ideally we should have remove/delete method for all our model-related classes and do cascade calls for all the children. To avoid those extra checks in `delete()/ remove()` we need to have something like factory that will instantiate from corresponding classes in runtime depending on the user's parameters. But I suppose it is big enough to become  separate jira. WDYT @michalkurka

Reply:
`cleanUpModelsCVPreds` is a bit different indeed: it's cleaned before finishing the AutoML run because we don't need it anymore (user didn't ask access to those predictions), but the AutoML object itself and all the attached models and metrics are still accessible.

the `delete/remove` logic is triggered either in tests, or when user explicitly deletes an object from the cluster using `h2o.delete(key)` or `h2o.rm(key)` (Py vs. R, didn't double check exactly).

Otherwise, agree of course about your OO/checks comments.
maybe for consistency, we should have `Model#remove_impl` call all of `deleteCrossValidationXXX` methods for now, incl. one for fold assignments... as I believe those CV models and predictions still leak today id user runs AutoML with the flags to keep them enabled, and then if he only deletes the models manually...